### PR TITLE
Add rapid router

### PR DESCRIPTION
**Pull request by SKairinos**
This change is [Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318)

### CODE REVIEW

File: backend/rapid_router/serializers/level.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    locked_for_class = LockedForClassField()
+
+    class Meta(BaseLevelSerializer.Meta):
+        fields = [*BaseLevelSerializer.Meta.fields, "locked_for_class"]
+        list_serializer_class = LockLevelListSerializer
+
+
+class ListLevelSerializer(BaseLevelSerializer):
+    class Meta(BaseLevelSerializer.Meta):
+        fields = [*BaseLevelSerializer.Meta.fields, "name", "locked_for_class"]
```
Comment (SKairinos) [inline backend/rapid_router/serializers/level.py:60]: return class access_code, not class id

File: backend/rapid_router/serializers/level_test.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+        self.level_1 = Level.objects.get(pk=1)
+        self.level_2 = Level.objects.get(pk=2)
+
+    # TODO: test this once access_code is changed to id in new schema.
+    # def test_update_many(self):
+    #     """Locks the levels for a specified class."""
+    #     levels = [self.level_1, self.level_2]
+
+    #     self.assert_update_many(
```
Comment (SKairinos) [inline backend/rapid_router/serializers/level_test.py:31]: write test without helper for now

File: backend/rapid_router/views/level_test.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        self.client.bulk_update(
+            models=levels,
+            data=[
+                {"locked_for_class": [klass.access_code]}
+                for _ in range(len(levels))
+            ],
+            action="lock",
+        )
+
+    def test_list(self):
```
Comment (SKairinos) [inline backend/rapid_router/views/level_test.py:150]: test that list can be filtered to only default levels

Comment (SKairinos) [reply backend/rapid_router/views/level_test.py:150]: Done

File: backend/rapid_router/serializers/level.py (hunk 1 above)
Comment (SKairinos) [reply backend/rapid_router/serializers/level.py:60]: Done

File: backend/rapid_router/serializers/level_test.py (hunk 2 above)
Comment (SKairinos) [reply backend/rapid_router/serializers/level_test.py:31]: Done.

**Review by faucomte97 — CHANGES_REQUESTED** [review]
> Reviewed 51 of 54 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318)* status: all files reviewed, 6 unresolved discussions (waiting on @SKairinos)

___
*[`backend/api/views/auth_factor.py` line 14 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBuwSK9cLp95LEccvw:-NvBuwSK9cLp95LEccvx:b2ni78k) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/api/views/auth_factor.py#L14)):*
> ```Python
> 
> 
> # pylint: disable-next=missing-class-docstring,too-ma,ny-ancestors
> ```

lol
___
*[`backend/rapid_router/serializers/level.py` line 37 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBwIbaAydtk5dBtSvn:-NvBwIbaAydtk5dBtSvo:b-wf583r) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/serializers/level.py#L37)):*
> ```Python
>         for level, data in zip(instance, validated_data):
>             for access_code in data["locked_for_class"]:
>                 level.locked_for_class.add(classes[access_code])
> ```

It seems like you only ever "add" to `locked_for_class`.
I can't see where you "remove" from it (or something of that nature), to basically "unclock" a level.

In the old system, on each request we a) unlock all the levels then b) lock the ones that get passed into the request.

https://github.com/ocadotechnology/codeforlife-portal/blob/master/portal/views/teacher/teach.py#L360
___
*[`backend/rapid_router/serializers/level.py` line 38 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBwsuw7Dedkqldgtz-:-NvBwsuw7Dedkqldgtz0:btifgxh) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/serializers/level.py#L38)):*
> ```Python
>             for access_code in data["locked_for_class"]:
>                 level.locked_for_class.add(classes[access_code])
> ```

Please also add the code which keeps track of the daily activity of the level control requests.

https://github.com/ocadotechnology/codeforlife-portal/blob/master/portal/views/teacher/teach.py#L364
___
*[`backend/rapid_router/serializers/level_test.py` line 31 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#gh-1560747937:-NvBxS_b7OeZi0RXFCRz:b-qjgyq4) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/736aad0cb32b14ed414e623e7727662af23d02b1/backend/rapid_router/serializers/level_test.py#L31)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I can't see the change?
___
*[`backend/rapid_router/serializers/level_test.py` line 41 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBxMUC9THB9FO3OlWY:-NvBxMUC9THB9FO3OlWZ:b-cvjxex) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/serializers/level_test.py#L41)):*
> ```Python
>     #     )
> 
>     def test_update_many(self):
> ```

Might be worth adding something here to test for unlocking a level
___
*[`backend/rapid_router/views/level_test.py` line 170 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBzTlI969SZCSSqS03:-NvBzTlI969SZCSSqS04:b-7z1zv1) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/views/level_test.py#L170)):*
> ```Python
>         self.client.list(
>             models=Level.objects.filter(default=True),
>             filters={"default": "true"},
> ```

Am a bit confused about this.
Can any field be passed in here? If so, why do we need the LevelFilter you made above?

**Review by faucomte97 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318)* status: all files reviewed, 7 unresolved discussions (waiting on @SKairinos)

*[a discussion](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBzhcH35Tpt3uiPt8Q:-NvBzhcH35Tpt3uiPt8R:b-xwkny6) (no related file):*
Install new PP version

___

**Review by SKairinos — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318)* status: 51 of 54 files reviewed, 7 unresolved discussions (waiting on @faucomte97)

___
*[`backend/api/views/auth_factor.py` line 14 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBuwSK9cLp95LEccvw:-NvCDCfm3Jv1cpqHP4TH:b-xybevs) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/api/views/auth_factor.py#L14)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

lol
</blockquote></details>

haha. good spot. fixed.
___
*[`backend/rapid_router/serializers/level.py` line 37 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBwIbaAydtk5dBtSvn:-NvCDkdtAs8aotIRhV-7:b-6gsyhn) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/serializers/level.py#L37)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

It seems like you only ever "add" to `locked_for_class`.
I can't see where you "remove" from it (or something of that nature), to basically "unclock" a level.

In the old system, on each request we a) unlock all the levels then b) lock the ones that get passed into the request.

https://github.com/ocadotechnology/codeforlife-portal/blob/master/portal/views/teacher/teach.py#L360
</blockquote></details>

Yes. Good spot! Thank you. Fixed.
___
*[`backend/rapid_router/serializers/level.py` line 38 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBwsuw7Dedkqldgtz-:-NvCF1gK0c9ffCOxDPcZ:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/serializers/level.py#L38)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

Please also add the code which keeps track of the daily activity of the level control requests.

https://github.com/ocadotechnology/codeforlife-portal/blob/master/portal/views/teacher/teach.py#L364
</blockquote></details>

Done.
___
*[`backend/rapid_router/serializers/level_test.py` line 31 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#gh-1560747937:-NvCF4IoD2v9Z0p5NDl3:b1mz9h2) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/736aad0cb32b14ed414e623e7727662af23d02b1/backend/rapid_router/serializers/level_test.py#L31)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

I can't see the change?
</blockquote></details>

it's the test below
___
*[`backend/rapid_router/serializers/level_test.py` line 41 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBxMUC9THB9FO3OlWY:-NvCFuFs3_0603SURasJ:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/serializers/level_test.py#L41)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

Might be worth adding something here to test for unlocking a level
</blockquote></details>

Done.
___
*[`backend/rapid_router/views/level_test.py` line 170 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvBzTlI969SZCSSqS03:-NvCFzOH9aTyKhTiCEfU:b-rh8dwe) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/b600cb6fa423bcf6f077a88d118f1cc78bf73157/backend/rapid_router/views/level_test.py#L170)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

Am a bit confused about this.
Can any field be passed in here? If so, why do we need the LevelFilter you made above?
</blockquote></details>

Only fields in the level's filter set should be passed here.

**Review by faucomte97 — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318)* status: all files reviewed, 4 unresolved discussions (waiting on @SKairinos)

___
*[`backend/rapid_router/serializers/level.py` line 40 at r4](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvCQZSQ9xrhD-ypI0z9:-NvCQZSQ9xrhD-ypI0zA:b1s0284) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f4ec762936b699d375c1c33d203989866770bc42/backend/rapid_router/serializers/level.py#L40)):*
> ```Python
>         }
> 
>         for klass in classes.values():
> ```

seems like we're allowing to lock levels for multiple classes in the same request? I noticed it earlier but didn't think to comment. Just double checking we do want this? Currently we can only lock levels for one class at a time.
___
*[`backend/rapid_router/serializers/level.py` line 80 at r4](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvCRCZH9t36LujySkui:-NvCRCZH9t36LujySkuj:b-rqyhaf) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f4ec762936b699d375c1c33d203989866770bc42/backend/rapid_router/serializers/level.py#L80)):*
> ```Python
>             if not queryset.filter(teacher__school=teacher.school).exists():
>                 raise serializers.ValidationError(
>                     "Class is not in the teacher's school.",
> ```

Why do we need this, doesn't the queryset in the view already check this?
___
*[`backend/rapid_router/serializers/level.py` line 89 at r4](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-NvCQr-k5WB0EBlB3xf9:-NvCQr-lCTJua0aF59Pr:b-qaz9j9) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f4ec762936b699d375c1c33d203989866770bc42/backend/rapid_router/serializers/level.py#L89)):*
> ```Python
>             ):
>                 raise serializers.ValidationError(
>                     "Teacher is not an admin or the class teacher.",
> ```

Why do we need this, don't the permissions in the view already check this? I'm asking because we don't check for this in the validate for any of the other actions, right?

**Review by faucomte97 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/318#-:-NvCSqO2E3wQy6thFk1r:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @SKairinos)

**Pull request merged**